src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -76,9 +87,72 @@ internal abstract class JsonPropertyInfo
         internal virtual void GetPolicies(JsonSerializerOptions options)
         {
             DetermineSerializationCapabilities(options);
+            DeterminePropertyName(options);
             IgnoreNullValues = options.IgnoreNullValues;
         }
 
+        private void DeterminePropertyName(JsonSerializerOptions options)
+        {
+            if (PropertyInfo != null)
+            {
+                JsonNameAttribute nameAttribute = GetAttribute<JsonNameAttribute>();
+                if (nameAttribute != null)
+                {
+                    NameAsString = nameAttribute.Name;
+                }
+                else if (options.PropertyNamingPolicy != null)
+                {
+                    NameAsString = options.PropertyNamingPolicy.CallConvertName(PropertyInfo.Name);
+                }
+                else
+                {
+                    NameAsString = PropertyInfo.Name;
+                }
+
+                // This is an exception case handled higher.
+                if (NameAsString == null)
+                {
+                    return;
+                }
+
+                // At this point propertyName is valid UTF16, so just call the simple UTF16->UTF8 encoder.
+                _name = Encoding.UTF8.GetBytes(NameAsString);
+
+                // Set the compare name.
+                if (options.PropertyNameCaseInsensitive)
```

Comment:
So does setting `PropertyNameCaseInsensitive` to true make all comparisons insensitive? So if someone had the class below and tried to deserialize with `PropertyNameCaseInsensitive` enabled, would the serializer throw an error complaining that there are duplicate properties?

```cs
public class Account
{
    public bool IsEnabled { get; set; }
    public bool isEnabled { get; set; }
}
```

The way Json.NET does case-insensitivity is it will look for an exact match first: `{ 'isEnabled': true }` would deserialized into `isEnabled`, then fallback to case-insensitive: `{ 'iSeNaBlEd': true }` would deserialized into `IsEnabled` because it is the first match based on property order.

Reply:
Yes the current semantics is that would throw during the caching phase which is the first (de)serialize for that type. The rationale is that it is ambiguous (at the high level) plus it is faster to just assume one check. 

@rynowak thoughts on this? Do we want to "probe" for exact match or throw?

Reply:
I think throwing for this case is fine. I can't understand why you'd want the serializer to tolerate duplicate properties

Reply:
Also Json.NET is case-insensitive by default so the double-lookup makes more sense there.

Reply:
My approach with Json.NET has been "serialize all the things". If you want to be more prescriptive then that's fine. Just letting you know of the options :smile: